pkg/didcomm/crypto/legacy/authcrypt/authcrypt.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package authcrypt
+
+import (
+	"crypto/sha512"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/agl/ed25519/extra25519"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/ed25519"
+	"golang.org/x/crypto/nacl/box"
+)
+
+type privateEd25519 [ed25519.PrivateKeySize]byte
+type publicEd25519 [ed25519.PublicKeySize]byte
+
+type keyPairEd25519 struct {
+	priv *privateEd25519
+	pub  *publicEd25519
+}
+
+// CurveKeySize is the size of public and private Curve25519 keys in bytes
+const CurveKeySize int = 32
+
+type privateCurve25519 [CurveKeySize]byte
+type publicCurve25519 [CurveKeySize]byte
+
+type keyPairCurve25519 struct {
+	priv *privateCurve25519
+	pub  *publicCurve25519
+}
+
+// contentEncryption represents a content encryption algorithm.
+// TODO: refactor together with jwe ContentEncryption
+type contentEncryption string
+
+// C20P Chacha20Poly1035 algorithm
+const C20P = contentEncryption("C20P") // Chacha20 encryption + Poly1035 authenticator cipher (96 bits nonce)
+
+// XC20P XChacha20Poly1035 algorithm
+const XC20P = contentEncryption("XC20P") // XChacha20 encryption + Poly1035 authenticator cipher (192 bits nonce)
+
+// Crypter represents an Authcrypt Encrypter (Decrypter) that outputs/reads legacy Aries envelopes
+type Crypter struct {
+	sender     keyPairEd25519
+	recipients []*publicEd25519
+	randSource io.Reader
+}
+
+// New will create a Crypter that encrypts messages using the legacy Aries format
+// Note: legacy crypter does not support XChacha20Poly1035 (XC20P), only Chacha20Poly1035 (C20P)
+func New(sender keyPairEd25519, recipients []*publicEd25519, randSource io.Reader) (*Crypter, error) { // nolint: lll
```

Comment:
rather than passing in a randSource instance to New (it's supposed to be hidden from the caller), use a variable randReader that you can update/mock in the tests then revert it back to crypt.Rand at the end of the test

Reply:
Ref payload from python decrypted in Go will come with Decrypter #294 
Rand source comment has been addressed in latest push